### ISSUE TITLE
Bug: Inconsistent accessibility

### ISSUE THREAD

**Issue by WhitWaldo** [postsharp/Metalama#162]
I'm introducing a dictionary as a field on a target type and per the introduction logic, I'm introducing it as `protected` accessibility. I'm then seeking to introduce a method which also has a `protected` accessibility value set in the introduction logic. I'm passing this dictionary into the method as one of its parameters with a `[CompileTime]` attribute, but they should have a consistent accessibility value.

Any guesses then, as to why I'm getting the error "CS0051	Inconsistent accessibility: parameter type 'Dictionary<string, List<IndexMetadata>>' is less accessible than method 'EntityStoreAttribute.__ProcessDictionaryClearedNotification_802981f71e2727dc(ITemplateSyntaxFactory, Dictionary<string, List<IndexMetadata>>, IFieldOrProperty, ExpressionSyntax, TemplateTypeArgument)'

Thank you!

**Comment by PostSharpBot** [comment]
Hello @WhitWaldo, thank you for submitting this issue. We will try to get back to you as soon as possible.
Note to the PostSharp team, this ticket is being tracked in our dashboard under ID TP-33369.

**Comment by svick** [comment]
Internally, Metalama currently creates a `public` method in the aspect type for each template (with parameters corresponding to, among other things, `[CompileTime]` parameters of the template). I think the error message indicates that your `List` type is less accessible than what's required for that, which means making your `List` more accessible should fix the error.

I'll look into improving the experience here.

**Comment by WhitWaldo** [comment]
I'm looking at the output if the diff if I just leave the dictionary off the list of arguments and that doesn't jive with what I'm seeing. Per the IMethodBuilder overrides, all the methods are created as `protected override void` and the fields referenced in the dictionary are also created as `protected`, so they should have the same visibility as one another.

Again, the dictionary itself is a compile-time only variable, so it shouldn't much matter what its visibility is at runtime as it doesn't exist then. It's just a wrapper for a collection of inserted fields so pass into methods at compile time. 

Any other guesses?

**Comment by svick** [comment]
@WhitWaldo The method I'm talking about is not visible in the diff and it's an implementation detail of Metalama.

For example, for this template:

```c#
[Template]
private dynamic? Template([CompileTime] int i)
{
    return i;
}
```

Metalama currently internally generates something like:

```c#
public SyntaxNode __Template_acd5d7ecfe556980(ITemplateSyntaxFactory templateSyntaxFactory, [CompileTime] int i)
{
    var __s1 = new List<StatementOrTrivia>();
    // return i;
    templateSyntaxFactory.AddStatement(
        __s1,
        templateSyntaxFactory.ReturnStatement(
            templateSyntaxFactory.RunTimeExpression(
                SyntaxFactory.LiteralExpression(
                    SyntaxKind.NumericLiteralExpression,
                    SyntaxFactory.Literal(i)))));
    return SyntaxFactory.Block(default, templateSyntaxFactory.ToStatementList(__s1));
}
```

Especially notice that the method is `public` even though the template is `private` and that it has the `[CompileTime]` parameter.

This method is then executed when you apply the template to generate the code that you see in the diff.

I will look into whether this method has to be public.

**Comment by WhitWaldo** [comment]
That does explain the funky method name in the error. Well, what does that mean for the dictionary I'm trying to pass into the method at compile-time then? It's not a field in the aspect, but just a variable created earlier in the `BuildAspect` method:
```cs
public override void BuildAspect(IAspectBuilder<INamedType> builder)
{
  // Setup bits
  Dictionary<string, List<TypeContainingIntroducedFields>> myDictionary = CreateMyDictionary();

  builder.Advice.IntroduceMethod(builder.Target, nameof(MyMethod), IntroductionScope.Target, OverrideStrategy.Override, b => {
    b.Accessibility = Accessibility.Protected;
  }, args: new {
    data = myDictionary
  }  
}

[Template]
private void MyMethod([CompileTime] Dictionary<string, List<TypeContainingIntroducedFields>>)
{
}
```

With this in mind then, I'm not entirely sure what I should do to make the dictionary more accessible... especially since I want it to remain protected in the actually generated class. Thoughts?

**Comment by svick** [comment]
I think a workaround would be to just make the template `public`. That shouldn't have any negative consequences.

**Comment by svick** [comment]
The workaround is to make the `TypeContainingIntroducedFields` type `public`. Which should not be a problem, since it won't affect the generated code.

**Comment by svick** [comment]
That's strange. Can you provide the code that reproduces the error? (Including all the relevant types.)

**Comment by WhitWaldo** [comment]
I'll get to work on that - do ya'll happen to have a non-public place I could submit my current repo to so as not to inadvertently and prematurely make it open source via support ticket? If not, I can work on trying to reproduce with a smaller demo project.

**Comment by gfraiteur** [comment]
You can create a private ticket on https://support.postsharp.net/ or ping us privately on Slack.

**Comment by svick** [comment]
@WhitWaldo Thanks for confirming. I have opened an internal ticket to improve the experience for such errors.

**Comment by prochan2** [comment]
This bug has been fixed in Metalama 2023.2.1-preview.
